### ISSUE TITLE
[Hold for payment 2024-02-07] [$500] MEDIUM: Fix Google Docs copy/paste of images

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#34306]
*Strategy:*
As a SaaS tool, we expect our customers to be using other best-of-breed SaaS products, especially Google Docs.  Having a fantastic integration for data exchange with Google Docs is a way to stand out.

*Problem:*
At one point in the past, we were able to just directly copy/paste charts and image from Google Docs, Google Sheets, Google Slides, etc -- straight into NewDot.  That doesn't work anymore.

*Solution:*
Add the ability to copy objects (not screenshots of objects, just the object itself) from Google Workspace into an Expensify chat, specifically:

* [x] Google Sheet charts
* [x] Google Slide images
* [x] Google Doc images

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0107c9e5dbbdb903ab</li>
        <li>Upwork Job ID: 1745258481290526720</li>
        <li>Last Price Increase: 2024-01-18</li>
<li>Automatic offers: </li>
<ul>
<li>ikevin127 | Contributor | 28116109</li>
</ul></li>
    </ul>
</details>

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
MEDIUM: Fix Google Docs copy/paste of images
### What is the root cause of that problem?

NA - Feature Request

### What changes do you think we should make in order to solve the problem?

When we copy an image from google, google will write the HTML content to the clipboard where it will includes source of the image copied etc.

```html
<meta charset='utf-8'>
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-ef70d940-7fff-9d33-5271-c83bd573c118"><span
        style="font-size:11pt;font-family:'Open Sans',sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;"><span
            style="border:none;display:inline-block;overflow:hidden;width:624px;height:451px;"><img
                src="https://lh7-us.googleusercontent.com/vt_1xrWmyFRWKb9zB-uFaougvcrbPSm5hdPaKUMEOtcJuHwtwFAJ_ubht3BAFoaDn59-S5AAJcaEPE8RCnsayNNjcDjhhXZTjWO1_6lgQ-EYbKCnpqiy0Xz5DFod1mOn4CDQHFtElQd8em7G-DqilBk"
                width="624" height="451" style="margin-left:0px;margin-top:0px;" /></span></span></b>
```

We're handling the pasting of HTML here https://github.com/Expensify/App/blob/8aec07c55df3f0fefc6bed76c991183cf6b97f7a/src/components/Composer/index.tsx#L226-L241 which will check if the pasted content is an emoji or not.

We should be adding a new condition here which will check for the source of the `img` tag i.e if url contains `googleusercontent` that means image is coming form google sources. 

> Any alternative is welcome, for this point validation is done using `URL` from JS

Once we have the url of the image, using `fetch` we get the image and create a new `File` object and pass it to `handlePasteFile` callback inside `Composer` file.

https://github.com/Expensify/App/blob/8aec07c55df3f0fefc6bed76c991183cf6b97f7a/src/components/Composer/index.tsx#L233-L238

In the above block of code, we should be adding the following code.

```jsx

try {
    // validate if the image source is a valid URL.
    const url = new URL(embeddedImages[0]?.src);
    // considering images from google, cause google will return the content as html
    if (embeddedImages[0].src && embeddedImages[0].src.includes('googleusercontent')) {
        handlePasteFileFromGoogleDoc(url.href);
        return;
    }
    // eslint-disable-next-line no-empty
} catch {}

```

```jsx
const handlePasteFileFromGoogleDoc = useCallback(
    (source: string) => {
        fetch(source, {method: 'GET'}).then((response) => {
            response.blob().then((blob) => {
                const fileType = response.headers.get('content-type') ?? 'image/png';
                const extension = fileType.split('/').pop() ?? 'png';
                // here name is `pasted-file` because the URL we're fetching isn't allowing to fetch the filename due to some CORS related issues..
                // if we want to fetch the file we should be creating a proxy URL where we take the image source and fetch and return via API we're using
                // which will solve issue.
                // Anyway I don't consider we need that cause, we're not caring for the image (or atleast I feel that way.)
                const blobToFile = new File([blob], `pasted-file.${extension}`, {type: fileType});
                onPasteFile(blobToFile);
            });
        });
    },
    [onPasteFile],
);

```




### What alternative solutions did you explore? (Optional)
NA

**Comment by ikevin127** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

The ability to directly copy/paste charts and images from Google Docs, Google Sheets, Google Slides straight into the chat.

## What is the root cause of that problem?

Missing functionality.

## What changes do you think we should make in order to solve the problem?

1. For copy & paste of files from Google Drive as attachments (drive.google.com/open?id= URL's):

(Requires Google API key)
Using Google Drive API on the front-end side we can check if a google drive file URL is pasted in the composer and add functionality that will extract the fileID and perform API call to download the file metadata from where we will do the following:
- get the file name and downloadLink
- perform final fetch where the actual file is downloaded
- transform file into blob using the original file name from the metadata
- pass the file to `onPasteFile(file)` in order to trigger the file management
- if fetch is successful the AttachmentModal will open with file preview and button to Send attachment to chat

2. For copy & paste of images from google docs and slides (googleusercontent.com URL's):

Check if clipboard data contains `googleusercontent.com` which is part of the image URL and if true:
- use `domparser` to extract embeddedImages from the parsed clipboard html data
- get the embeddedImage[0.src URL and fetch the file then transform response into blob
- pass the file to `onPasteFile(file)` in order to trigger the file management
- AttachmentModal will open with file preview and button to Send attachment to chat

3. For copy & paste of google sheet charts  / images (base64 source):

Check if clipboard data contains base64 image source and if true perform the folowing actions:
- use `domparser` to extract embeddedImages from the parsed clipboard html data
- extract base64ImageContent and mimeType then make it into a blob
- pass the file to `onPasteFile(file)` in order to trigger the file management
- AttachmentModal will open with file preview and button to Send attachment to chat

## Videos

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/051c6da7-4b6d-4e6f-beb3-d9e4caea7d12

</details>

**Comment by conorpendergrast** [comment]
@Santhosh-Sellavel Could you take a look at the two proposals so far and confirm if we can move ahead with either of them?

**Comment by Santhosh-Sellavel** [comment]
@conorpendergrast I'll try to get a hold of it by tomorrow.

**Comment by quinthar** [comment]
@Santhosh-Sellavel tomorrow came and went; can you give a new ETA for when you can handle this?

**Comment by Santhosh-Sellavel** [comment]
Sorry for the delay, everyone.

Both @b4s36t4 & @ikevin127's proposal looks good here.

But @b4s36t4's proposals work well for Google Docs & Google Slides, but it doesn't work for charts copied from Google Sheets.

@ikevin127 I like your proposal, looks good to me as it addresses charts case from Google Sheets case as well.

**Comment by b4s36t4** [comment]
@Santhosh-Sellavel as far as I see when I copy the charts from google sheets, It is coping the string (values) from the graph. Even in the google docs it's not working perfectly.

<img width="1728" alt="Screenshot 2024-01-22 at 11 28 27 PM" src="https://github.com/Expensify/App/assets/59088937/2a474f4c-fcea-4a4f-9cdb-4245d888d25e">
<img width="585" alt="Screenshot 2024-01-22 at 11 29 34 PM" src="https://github.com/Expensify/App/assets/59088937/d6a3436b-87e2-48b7-a795-5ca4d1e52081">

> Having the google docs API to copy data from google docs etc wouldn't good I feel, in terms of UX as well since that require api calling and pricing is also we need to check.

**Comment by b4s36t4** [comment]
To fix the graph issue, I'm working on a alternate solution which should solve the issue (but might require a package to be installed).

**Comment by b4s36t4** [comment]
Also, if you could share the chart you've used to test the sheets issue, that'd be helpful for me to validate my proposal. Since the other proposal is majorly same as my proposal :).

**Comment by Santhosh-Sellavel** [comment]
> @Santhosh-Sellavel as far as I see when I copy the charts from google sheets, It is coping the string (values) from the graph. Even in the google docs it's not working perfectly.

Those might be actually string values.

**Comment by b4s36t4** [comment]
<img width="428" alt="Screenshot 2024-01-22 at 11 47 40 PM" src="https://github.com/Expensify/App/assets/59088937/e8e152a3-99f6-45cb-b7f2-af089c3d76d1">

https://docs.google.com/spreadsheets/d/17N_v96fqMSMxBtynS8qlCkUPYevsStx7aw-uolUIuN4/edit#gid=0

Source of the copied chart.

**Comment by Santhosh-Sellavel** [comment]
There are different represent

> <img alt="Screenshot 2024-01-22 at 11 47 40 PM" width="428" src="https://private-user-images.githubusercontent.com/59088937/298682475-e8e152a3-99f6-45cb-b7f2-af089c3d76d1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5NDc3NzUsIm5iZiI6MTcwNTk0NzQ3NSwicGF0aCI6Ii81OTA4ODkzNy8yOTg2ODI0NzUtZThlMTUyYTMtOTlmNi00NWNiLWI3ZjItYWYwODljM2Q3NmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIyVDE4MTc1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzN2I1ZmRlNGNmMTA3Zjc1YTcwNDhjZjZiYjhhN2FkZDE1MTRkNWNhMjdjZTgyYmQ2MDBhNzdjZDM3NDA2NDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dOzBQHr97sqhusAAflgJB0LprO19HGJBSXy3cy4qX-Q">
> https://docs.google.com/spreadsheets/d/17N_v96fqMSMxBtynS8qlCkUPYevsStx7aw-uolUIuN4/edit#gid=0
> 
> Source of the copied chart.

That's a table representation, check for different templates involves actual charts like the following templates

<img width="235" alt="Screenshot 2024-01-22 at 11 50 44 PM" src="https://github.com/Expensify/App/assets/85645967/0089067c-dc43-48e8-bfe3-4a4c5a7e391b">
<img width="231" alt="Screenshot 2024-01-22 at 11 50 58 PM" src="https://github.com/Expensify/App/assets/85645967/d2b40771-808c-4837-ab25-aaeef9705de2">

**Comment by b4s36t4** [comment]
@aimane-chnaif I'm on it actually. Will be fixing by coming 24hrs. Sorry for the delay.

**Comment by ikevin127** [comment]
> Since the other proposal is majorly same as my proposal :)

@b4s36t4 Yeah, "majorly same" except for the 2 other cases you didn't account for. Thanks for the laugh tho 😅

**Comment by b4s36t4** [comment]
https://github.com/Expensify/App/issues/34306#issuecomment-1886189109


https://github.com/Expensify/App/assets/59088937/4a73c962-d91b-4741-8390-e6f11ee8ad39 - Demo after updating my proposal.

Handling copy/paste from google drive is worthless I feel.

* Need of API key.
* User need make the file public (because the API key we're using different not related to user's org or account)
* if not public, user have to share it to the email that we've created with.
* Private files, they don't want to share to public or outside of their known people?

cc: @Santhosh-Sellavel  Please review the updated proposal thanks :). Sorry for the ping.

**Comment by ikevin127** [comment]
Offer accepted!

I will have the PR open by early tomorrow.

**Comment by b4s36t4** [comment]
@mountiny requesting a re-review of proposal. No worries if nothing changed.

**Comment by mountiny** [comment]
I trust the c+ with their selection in this case, please do focus on completing already assigned tasks first 🙇

**Comment by ikevin127** [comment]
@Santhosh-Sellavel PR https://github.com/Expensify/App/pull/34971 is ready for review! 🚀

**Comment by b4s36t4** [comment]
haha, sorry if I was too intimidating, but the PR is literal copy of my proposal. Hoping this is taken care!!!!.

>PS:  I'm not expecting any sharing of bounty, but the efforts should be credited atleast.

**Comment by ikevin127** [comment]
Thank you for your contribution to this issue!

**Comment by mountiny** [comment]
Thanks to you both, it will go the other way around in future

**Comment by ikevin127** [comment]
Some manual testing using the external `googleusercontent.com` links (Docs / Slides images) did not pass on staging after deploy, I added more details in a PR comment here https://github.com/Expensify/App/pull/34971#issuecomment-1918697405.

**TLDR:** the link violates our Content Security Policy directive: "connect-src" because the `https://*.googleusercontent.com` is missing from the allowed links list, therefore the fetch fails.

I thought I should let the team handling this issue know about it in case they are not getting the PR notifications anymore.

cc @mountiny @Santhosh-Sellavel

**Comment by mountiny** [comment]
Created backend change to allow for this content-src https://github.com/Expensify/Cloudflare-Workers/pull/119

**Comment by quinthar** [comment]
Hi, why did @github-project-automation move to CRITICAL, I'm curious?

**Comment by quinthar** [comment]
Also, I think there's more work to be done here, though it's off to an awesome start!  I am able to copy/paste a chart from Google Sheet directly into NewDot, which is great.  However, this didn't work:

1. Open an empty Google Doc
2. Open any image from the internet (such as [this cat](https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.nationalgeographic.com%2Fanimals%2Fmammals%2Ffacts%2Fdomestic-cat&psig=AOvVaw0NZytJARx6wFGv2XmqgYC0&ust=1706859545771000&source=images&cd=vfe&opi=89978449&ved=0CBMQjRxqFwoTCIiNlbnRiYQDFQAAAAAdAAAAABAE))
3. Copy and paste that into the Google Doc, so now it's a Google Doc image object
4. Now select and copy that image *from Google Docs*
5. **Correct:** You should be able to paste it into a NewDot chat
6. **Actual:** When you try to paste it, nothing happens.

Reopening this issue to fix this.  Thanks for the great work on this so far!

**Comment by conorpendergrast** [comment]
I tested out Google Slides, and that also works! So it looks like it's just Google Docs that's remaining (I've ticked off Sheets and Slides in the OP)

**Comment by ikevin127** [comment]
Appreciate the amazing feedback! 🚀

I'm having some trouble understanding how come the Google Docs scenario mentioned here https://github.com/Expensify/App/issues/34306#issuecomment-1920691630 doesn't work 🧐

**Note**: I'm pretty sure it's not a Google Doc permission related issue as the images / objects copied to the clipboard cannot be restricted even if the Google Doc file is Restricted.

We implemented 2 types of clipboard paste for this feature so far:
- `base64` images / objects (charts) mostly found in Google Sheets
- `googleusercontent.com` links mostly found in Google Docs / Slides images

Since the scenario mentioned in the comment above matches the 2nd clipboard type, there shouldn't be any issue❓
Here's how it looks on my side following the steps:

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/50caf4db-0709-4617-86c2-6ebbee53edc3

</details>

cc @quinthar @conorpendergrast

**Comment by quinthar** [comment]
Wow, amazing video!

**Comment by quinthar** [comment]
I can't reproduce the issue anymore, not sure what I did.  Closing again.  Great work!!

**Comment by conorpendergrast** [comment]
Reopening to work out payment. Looks like the automation failed there, so I'll handle once the regression period is over (looks like that's 2024-02-14)

**Comment by conorpendergrast** [comment]
**Payouts due:**

- [x] Contributor: $500 @ikevin127 - [Contract](https://www.upwork.com/nx/wm/workroom/36101802/overview)
Contributor+: $500 @Santhosh-Sellavel - via Manual request

Upwork job is [here](https://www.upwork.com/jobs/~0107c9e5dbbdb903ab).

**Comment by Santhosh-Sellavel** [comment]
Requested on ND

**Comment by JmillsExpensify** [comment]
$500 approved for @Santhosh-Sellavel based on summary above.
